schema-engine-tag/src/main/java/org/apache/iotdb/lsm/util/BloomFilter.java

```diff
@@ -0,0 +1,238 @@
+    this.hashFunctionSize = hashFunctionSize;
+    func = new HashFunction[hashFunctionSize];
+    for (int i = 0; i < hashFunctionSize; i++) {
+      func[i] = new HashFunction(size, SEEDS[i]);
+    }
+
+    bits = new BitSet(size);
+  }
+
+  /**
+   * get empty bloom filter
+   *
+   * @param errorPercent the tolerant percent of error of the bloom filter
+   * @param numOfString the number of string want to store in the bloom filter
+   * @return empty bloom
+   */
+  public static BloomFilter getEmptyBloomFilter(double errorPercent, int numOfString) {
+    errorPercent = Math.max(errorPercent, TSFileConfig.MIN_BLOOM_FILTER_ERROR_RATE);
+    errorPercent = Math.min(errorPercent, TSFileConfig.MAX_BLOOM_FILTER_ERROR_RATE);
+
+    double ln2 = Math.log(2);
+    int size = (int) (-numOfString * Math.log(errorPercent) / ln2 / ln2) + 1;
+    int hashFunctionSize = (int) (-Math.log(errorPercent) / ln2) + 1;
+    return new BloomFilter(
+        Math.max(MINIMAL_SIZE, size), Math.min(MAXIMAL_HASH_FUNCTION_SIZE, hashFunctionSize));
+  }
+
+  /**
+   * build bloom filter by bytes
+   *
+   * @param bytes bytes of bits
+   * @return bloom filter
+   */
+  public static BloomFilter buildBloomFilter(byte[] bytes, int size, int hashFunctionSize) {
+    return new BloomFilter(bytes, size, Math.min(MAXIMAL_HASH_FUNCTION_SIZE, hashFunctionSize));
+  }
+
+  public int getHashFunctionSize() {
+    return hashFunctionSize;
+  }
+
+  public int getSize() {
+    return size;
+  }
+
+  public void setSize(int size) {
+    this.size = size;
+  }
+
+  public void add(String value) {
+    for (HashFunction f : func) {
+      bits.set(f.hash(value), true);
+    }
+  }
+
+  public boolean contains(String value) {
+    if (value == null) {
+      return false;
+    }
+    boolean ret = true;
+    int index = 0;
+    while (ret && index < hashFunctionSize) {
+      ret = bits.get(func[index++].hash(value));
+    }
+
+    return ret;
+  }
+
+  public int getBitCount() {
+    int res = 0;
+    for (int i = 0; i < size; i++) {
+      res += bits.get(i) ? 1 : 0;
+    }
+
+    return res;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    BloomFilter that = (BloomFilter) o;
+    return size == that.size
+        && hashFunctionSize == that.hashFunctionSize
+        && Objects.equals(bits, that.bits)
+        && Arrays.equals(func, that.func);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(size, hashFunctionSize, bits, func);
+  }
+
+  private class HashFunction {
+
+    private int cap;
+    private int seed;
+
+    HashFunction(int cap, int seed) {
+      this.cap = cap;
+      this.seed = seed;
+    }
+
+    public int hash(String value) {
+      int res = Murmur128Hash.hash(value, seed);
+      if (res == Integer.MIN_VALUE) {
+        res = 0;
+      }
+
+      return Math.abs(res) % cap;
+    }
```

Comment:
This method can be refactored specially for our tag model. 

We can compute the hash code of tag key and tag value separately and combine them by adding together or some way else to act as the final hash code of key+value. 

When batch processing tag values, the internal computing result of tag key can be cached and reused.

Reply:
done